### PR TITLE
Gkd trainer

### PR DESCRIPTION
**Pull request by kashif**
https://arxiv.org/pdf/2306.13649

### CODE REVIEW

File: trl/trainer/gkd_trainer.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        Return:
+            `torch.Tensor`: The tensor with training loss on this batch.
+        """
+        if random.random() >= self.lmbda:
+            # On-policy: Generate outputs from the student model with temperature self.temperature
+            with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
+                generation_config = GenerationConfig(
+                    max_new_tokens=self.args.max_new_tokens_response,
+                    temperature=self.temperature,
+                    do_sample=True,
```
Comment (lewtun) [inline trl/trainer/gkd_trainer.py:176]: If I'm not mistaken, we need to override the default value for top_k to do pure sampling: https://huggingface.co/docs/transformers/v4.44.0/en/main_classes/text_generation#transformers.GenerationConfig

```suggestion
                    do_sample=True,
                    top_k=0,
```

File: examples/scripts/gkd.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        quantization_config=quantization_config,
+    )
+    training_args.model_init_kwargs = model_kwargs
+    tokenizer = AutoTokenizer.from_pretrained(
+        model_config.model_name_or_path,
+        trust_remote_code=model_config.trust_remote_code,
+        use_fast=True,
+    )
+    tokenizer.pad_token = tokenizer.eos_token
+    tokenizer.model_max_length = 1024
```
[the remark this answers is not in the document]
Comment (kashif) [reply examples/scripts/gkd.py:122]: some tokenizers i was testing with had undefined or huge model_max_lenghts... I will remove

File: examples/scripts/gkd.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+# limitations under the License.
+"""
+# Full training:
+python examples/scripts/gkd.py \
+    --model_name_or_path="Qwen/Qwen2-0.5B-Instruct" \
+    --teacher_model_name_or_path="Qwen/Qwen2-1.5B-Instruct" \
+    --dataset_name="andito/chatbot_arena_completions" \
+    --report_to="wandb" \
+    --learning_rate=2e-5 \
+    --per_device_train_batch_size=64 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:23]:
```suggestion
    --per_device_train_batch_size=16 \
```

File: examples/scripts/gkd.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    --gradient_accumulation_steps=16 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:24]:
```suggestion
    --gradient_accumulation_steps=2 \
```

File: examples/scripts/gkd.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    --output_dir="gkd-model" \
+    --logging_steps=1 \
+    --num_train_epochs=3 \
+    --max_steps=-1 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:28]:
```suggestion
```

File: examples/scripts/gkd.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    --gradient_checkpointing
+
+# LoRA:
+python examples/scripts/gkd.py \
+    --model_name_or_path="Qwen/Qwen2-0.5B-Instruct" \
+    --teacher_model_name_or_path="Qwen/Qwen2-1.5B-Instruct" \
+    --dataset_name="andito/chatbot_arena_completions" \
+    --report_to="wandb" \
+    --learning_rate=2e-4 \
+    --per_device_train_batch_size=64 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:39]:
```suggestion
    --per_device_train_batch_size=16 \
```

File: examples/scripts/gkd.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    --gradient_accumulation_steps=16 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:40]:
```suggestion
    --gradient_accumulation_steps=2 \
```

File: examples/scripts/gkd.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    --output_dir="gkd-model" \
+    --logging_steps=1 \
+    --num_train_epochs=3 \
+    --max_steps=-1 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:44]:
```suggestion
```

**Review by lewtun — COMMENTED** [review]
FYI I added a few tweaks to the example scripts so they can fit on a single GPU

File: trl/trainer/utils.py [hunk 9]
```diff
@@ -214,6 +214,89 @@ def torch_call(self, examples: List[Union[List[int], Any, Dict[str, Any]]]) -> D
...
+
+    tokenizer: PreTrainedTokenizerBase
+    ignore_index: int = -100
+    max_length: int = None
+
+    def __post_init__(self):
+        if self.tokenizer.pad_token_id is None:
+            raise ValueError("The tokenizer does not have a pad token. Please set `pad_token_id` in the tokenizer.")
+        if self.max_length is None:
+            self.max_length = self.tokenizer.model_max_length
```
Comment (lewtun) [inline trl/trainer/utils.py:231]: Some models have `model_max_length` set to the largest Python `int` value, so for those cases we need a sensible default value + a warning

File: trl/trainer/gkd_trainer.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+                The model to train.
+            inputs (`Dict[str, Union[torch.Tensor, Any]]`):
+                The inputs and targets of the model.
+
+                The dictionary will be unpacked before being fed to the model. Most models expect the targets under the
+                argument `labels`. Check your model's documentation for all accepted arguments.
+
+        Return:
+            `torch.Tensor`: The tensor with training loss on this batch.
+        """
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py:180]: You can remove this if it's the same as `Trainer.training_step`

File: tests/test_gkd_trainer.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (qgallouedec) [inline tests/test_gkd_trainer.py:1]:
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

File: tests/test_gkd_trainer.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                model=self.model_id,
+                teacher_model=self.model_id,
+                args=training_args,
+                train_dataset=self.train_dataset,
+                eval_dataset=self.eval_dataset,
+            )
+
+            trainer.train()
+
+            assert trainer.state.log_history[(-1)]["train_loss"] is not None
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/test_gkd_trainer.py:148]:
```suggestion
            self.assertIsNotNone(trainer.state.log_history[(-1)]["train_loss"])
```

File: trl/trainer/gkd_config.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        teacher_model_init_kwargs (:obj:`Dict`, `optional`, defaults to `None`):
+            The initialization kwargs to use when creating the teacher model. If `None`, the teacher model will be
+            instantiated with the default initialization kwargs.
+        disable_dropout (:obj:`bool`, `optional`, defaults to `True`):
+            Whether to disable dropout in the model being trained.
+    """
+
+    temperature: float = 1.0
+    lmbda: float = 0.5
+    beta: float = 0.5
```
Comment (qgallouedec) [inline trl/trainer/gkd_config.py:46]: Higher than usual. Is it the recommended value from the paper?

File: trl/trainer/gkd_trainer.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+
+class GKDTrainer(SFTTrainer):
+    _tag_names = ["trl", "gkd"]
+
+    def __init__(
+        self,
+        teacher_model: Union[PreTrainedModel, nn.Module, str],
+        args: Optional[GKDConfig] = None,
+        *sft_args,
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py:45]: personal preference: I'd have them explicitly here

File: tests/test_gkd_trainer.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+def formatting_prompts_func_batched(example):
+    output_text = []
+    for i, question in enumerate(example["question"]):
+        text = f"### Question: {question}\n ### Answer: {example['answer'][i]}"
+        output_text.append(text)
+    return output_text
+
+
+if is_peft_available():
+    pass
```
Comment (qgallouedec) [inline tests/test_gkd_trainer.py:42]:
```suggestion
```

**Comment by kashif** [comment]
@lewtun added student generations as well in the `gkd.py` script

File: docs/source/gkd_trainer.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
+# Generalized Knowledge Distillation Trainer
+
+TRL supports Generalized Knowledge Distillation (GKD), which is a new approach to distill a student model from a larger teacher model, as described in the paper [On-Policy Distillation of Language Models: Learning from Self-Generated Mistakes](https://huggingface.co/papers/2306.13649).  The key aspects of GKD are:
```
Comment (qgallouedec) [inline docs/source/gkd_trainer.md:3]:
```suggestion
TRL supports Generalized Knowledge Distillation (GKD), which is a new approach to distill a student model from a larger teacher model, as described in the paper [On-Policy Distillation of Language Models: Learning from Self-Generated Mistakes](https://huggingface.co/papers/2306.13649). The key aspects of GKD are:
```

File: docs/source/gkd_trainer.md [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+1. It addresses the train-inference distribution mismatch in auto-regressive sequence models by training the student model on its self-generated output sequences.
+2. GKD allows flexibility in choosing different divergence measures between student and teacher models via the generalized Jensen-Shannon Divergence (JSD), which can be useful when the student lacks the capacity to fully mimic the teacher.
+
+The GKD Trainer is a wrapper around the [`SFTTrainer`] class that takes in a teacher model argument. It needs two parameters to be set via the [`GKDConfig`]:
+* `lmbda`:  controls the student data fraction, i.e., the proportion of on-policy student-generated outputs. When `lmbda=0`, the loss reduces to supervised JSD where the student is trained with the token-level probabilities of the teacher. When `lmbda=1`, the loss reduces to on-policy JSD, where the student generates output sequences and token-specific feedback on these sequences from the teacher. For values in between [0,1] it is random between the two based on the `lmbda` value for each batch.
```
Comment (qgallouedec) [inline docs/source/gkd_trainer.md:8]:
```suggestion
* `lmbda`:  controls the student data fraction, i.e., the proportion of on-policy student-generated outputs. When `lmbda=0.0`, the loss reduces to supervised JSD where the student is trained with the token-level probabilities of the teacher. When `lmbda=1.0`, the loss reduces to on-policy JSD, where the student generates output sequences and token-specific feedback on these sequences from the teacher. For values in between [0,1] it is random between the two based on the `lmbda` value for each batch.
```

File: docs/source/gkd_trainer.md [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+* `beta`: controls the interpolation in the generalized Jensen-Shannon Divergence.  When `beta=0` the loss approximates forward KL divergence, while for `beta=1` the loss approximates reverse KL divergence. For values in between [0,1] it interpolates between the two.
```
Comment (qgallouedec) [inline docs/source/gkd_trainer.md:9]:
```suggestion
* `beta`: controls the interpolation in the generalized Jensen-Shannon Divergence.  When `beta=0.0` the loss approximates forward KL divergence, while for `beta=1.0` the loss approximates reverse KL divergence. For values in between [0, 1] it interpolates between the two.
```

File: examples/scripts/gkd.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+    try:
+        eval_dataset = raw_datasets[args.dataset_test_split]
+        prompts = eval_dataset["messages"][:8]
+    except KeyError:
+        eval_dataset = None
+        prompts = train_dataset["messages"][:8]
+
+    # remove the last assistant message from the prompts messages and then apply chat template to the prompts
+    prompts = [prompts[i][:-1] for i in range(len(prompts))]
+    prompts = [tokenizer.apply_chat_template(prompt, tokenize=False) for prompt in prompts]
```
Comment (qgallouedec) [inline examples/scripts/gkd.py:132]:
```suggestion
    prompts = [tokenizer.apply_chat_template(prompt, tokenize=False, add_generation_prompt=True) for prompt in prompts]
```

File: examples/scripts/gkd.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    --gradient_accumulation_steps=2 \
+    --output_dir="gkd-model" \
+    --logging_steps=1 \
+    --num_train_epochs=3 \
+    --push_to_hub \
+    --gradient_checkpointing
+
+# LoRA:
+python examples/scripts/gkd.py \
+    --model_name_or_path="Qwen/Qwen2-0.5B-Instruct" \
```
Comment (lewtun) [inline examples/scripts/gkd.py:33]:
```suggestion
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
```

File: examples/scripts/gkd.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    --teacher_model_name_or_path="Qwen/Qwen2-1.5B-Instruct" \
```
Comment (lewtun) [inline examples/scripts/gkd.py:34]:
```suggestion
    --teacher_model_name_or_path Qwen/Qwen2-1.5B-Instruct \
```

File: examples/scripts/gkd.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+    --dataset_name="andito/chatbot_arena_completions" \
```
Comment (lewtun) [inline examples/scripts/gkd.py:35]:
```suggestion
    --dataset_name andito/chatbot_arena_completions \
```

File: examples/scripts/gkd.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+    --report_to="wandb" \
```
Comment (lewtun) [inline examples/scripts/gkd.py:36]:
```suggestion
```

File: examples/scripts/gkd.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+    --learning_rate=2e-4 \
+    --per_device_train_batch_size=16 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:38]:
```suggestion
    --per_device_train_batch_size 4 \
```

File: examples/scripts/gkd.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+    --push_to_hub \
+    --gradient_checkpointing
+
+# LoRA:
+python examples/scripts/gkd.py \
+    --model_name_or_path="Qwen/Qwen2-0.5B-Instruct" \
+    --teacher_model_name_or_path="Qwen/Qwen2-1.5B-Instruct" \
+    --dataset_name="andito/chatbot_arena_completions" \
+    --report_to="wandb" \
+    --learning_rate=2e-4 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:37]:
```suggestion
    --learning_rate 2e-4 \
```

File: examples/scripts/gkd.py [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+    --per_device_train_batch_size=16 \
+    --gradient_accumulation_steps=2 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:39]:
```suggestion
    --gradient_accumulation_steps 8 \
```

File: examples/scripts/gkd.py [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+    --output_dir="gkd-model" \
```
Comment (lewtun) [inline examples/scripts/gkd.py:40]:
```suggestion
    --output_dir gkd-model \
```

File: examples/scripts/gkd.py [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+    --logging_steps=1 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:41]:
```suggestion
    --logging_steps 10 \
```

File: examples/scripts/gkd.py [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+    --num_train_epochs=3 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:42]:
```suggestion
    --num_train_epochs 1 \
```

File: examples/scripts/gkd.py [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+    --push_to_hub \
+    --gradient_checkpointing \
+    --use_peft \
+    --lora_r=64 \
```
Comment (lewtun) [inline examples/scripts/gkd.py:46]:
```suggestion
    --lora_r 64 \
```

File: examples/scripts/gkd.py [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+    --lora_alpha=16
```
Comment (lewtun) [inline examples/scripts/gkd.py:47]:
```suggestion
    --lora_alpha 16
```

File: examples/scripts/gkd.py [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+        use_cache=False if training_args.gradient_checkpointing else True,
+        device_map=get_kbit_device_map() if quantization_config is not None else None,
+        quantization_config=quantization_config,
+    )
+    training_args.model_init_kwargs = model_kwargs
+
+    tokenizer = AutoTokenizer.from_pretrained(
+        model_config.model_name_or_path,
+        trust_remote_code=model_config.trust_remote_code,
+        use_fast=True,
```
Comment (lewtun) [inline examples/scripts/gkd.py:113]: Needed for correct generations:

```suggestion
        padding_side="left",
```

**Review by lewtun — COMMENTED** [review]
Looking good! Added some nits to the example and a request to align the docs with the online DPO format

**Comment by lewtun** [comment]
The current implementation seems to fail for Gemma models with 

```
RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
[rank0]: Traceback (most recent call last):
[rank0]:   File "/fsx/lewis/git/hf/trl/examples/scripts/gkd.py", line 160, in <module>
[rank0]:     trainer.train()
[rank0]:   File "/fsx/lewis/git/hf/trl/trl/trainer/sft_trainer.py", line 450, in train
[rank0]:     output = super().train(*args, **kwargs)
[rank0]:   File "/fsx/lewis/miniconda3/envs/trl/lib/python3.10/site-packages/transformers/trainer.py", line 1938, in train
[rank0]:     return inner_training_loop(
[rank0]:   File "/fsx/lewis/miniconda3/envs/trl/lib/python3.10/site-packages/transformers/trainer.py", line 2279, in _inner_training_loop
[rank0]:     tr_loss_step = self.training_step(model, inputs)
[rank0]:   File "/fsx/lewis/git/hf/trl/trl/trainer/gkd_trainer.py", line 211, in training_step
[rank0]:     new_input_ids, new_attention_mask = self.generate_on_policy_outputs(
[rank0]:   File "/fsx/lewis/git/hf/trl/trl/trainer/gkd_trainer.py", line 182, in generate_on_policy_outputs
[rank0]:     generated_outputs = model.generate(
[rank0]:   File "/fsx/lewis/miniconda3/envs/trl/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/fsx/lewis/miniconda3/envs/trl/lib/python3.10/site-packages/transformers/generation/utils.py", line 2024, in generate
[rank0]:     result = self._sample(
[rank0]:   File "/fsx/lewis/miniconda3/envs/trl/lib/python3.10/site-packages/transformers/generation/utils.py", line 3020, in _sample
[rank0]:     next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
[rank0]: RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
```

Command to repro:

```shell
accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero2.yaml  examples/scripts/gkd.py     --model_name_or_path google/gemma-2-2b-it     --teacher_model_name_or_path google/gemma-2-9b-it     --dataset_name andito/chatbot_arena_completions      --learning_rate=5e-6     --per_device_train_batch_size=4     --gradient_accumulation_steps=1     --output_dir gkd-model     --logging_steps=5  --num_train_epochs=1 --gradient_checkpointing --bf16 --attn_implementation eager
```

File: docs/source/gkd_trainer.md [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+
+## Usage tips
+
+The GKD Trainer is a wrapper around the [`SFTTrainer`] class that takes in a teacher model argument. It needs two parameters to be set via the [`GKDConfig`] namely:
+* `lmbda`:  controls the student data fraction, i.e., the proportion of on-policy student-generated outputs. When `lmbda=0.0`, the loss reduces to supervised JSD where the student is trained with the token-level probabilities of the teacher. When `lmbda=1.0`, the loss reduces to on-policy JSD, where the student generates output sequences and token-specific feedback on these sequences from the teacher. For values in between [0, 1] it is random between the two based on the `lmbda` value for each batch.
+* `beta`: controls the interpolation in the generalized Jensen-Shannon Divergence.  When `beta=0.0` the loss approximates forward KL divergence, while for `beta=1.0` the loss approximates reverse KL divergence. For values in between [0, 1] it interpolates between the two.
+
+The authors find that on-policy data (high `lmbda`) performs better and the optimal `beta` varied depending on the task and evaluation method.
```
Comment (lewtun) [inline docs/source/gkd_trainer.md:24]:
```suggestion
The authors find that on-policy data (high `lmbda`) performs better and the optimal `beta` varied depending on the task and evaluation method.

The basic API is as follows:

```python
from datasets import Dataset
from trl import GKDConfig, GKDTrainer
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
)

NUM_DUMMY_SAMPLES = 100

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
# The model to optimise
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
# The teacher model to calculate the KL divergence against
teacher_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")

train_dataset = Dataset.from_dict(
    {
        "messages": [
            [
                {"role": "user", "content": "Hi, how are you?"},
                {"role": "assistant", "content": "I'm great thanks"},
            ]
        ]
        * NUM_DUMMY_SAMPLES
    }
)
eval_dataset = Dataset.from_dict(
    {
        "messages": [
            [
                {"role": "user", "content": "What colour is the sky?"},
                {"role": "assistant", "content": "The sky is blue"},
            ]
        ]
        * NUM_DUMMY_SAMPLES
    }
)

args = GKDConfig(output_dir="gkd-model", per_device_train_batch_size=1)
trainer = GKDTrainer(
    model=model,
    teacher_model=teacher_model,
    args=args,
    tokenizer=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
)
trainer.train()
```
```

**Review by lewtun — APPROVED** [review]
Thanks for ironing out the final kinks! I've tested Qwen and Gemma models and it seems to work well - we can always iterate a bit on the small details in follow up PRs.

I've left some final suggestions and have a question about the need to pass `sft_args` to the training vs just having a single set of `args`. Is it because the parent class won't accept the new params of the GKD config? If that's the case, then let's keep it as is and merge once you've incorporated the final changes

**Comment by lewtun** [comment]
Any idea why the doc build is failing @kashif ?

**Comment by kashif** [comment]
not sure... i was also checking

**Pull request merged**